Fix compilation without deprecated OpenSSL 1.1 APIs

neheb opened the pull request.

nikic commented:
```
/home/travis/build/php/php-src/ext/openssl/openssl.c: In function ‘zm_info_openssl’:
/home/travis/build/php/php-src/ext/openssl/openssl.c:1623:57: error: implicit declaration of function ‘OpenSSL_version’ [-Werror=implicit-function-declaration]
  php_info_print_table_row(2, "OpenSSL Library Version", OpenSSL_version(OPENSSL_VERSION));
                                                         ^
/home/travis/build/php/php-src/ext/openssl/openssl.c:1623:73: error: ‘OPENSSL_VERSION’ undeclared (first use in this function)
  php_info_print_table_row(2, "OpenSSL Library Version", OpenSSL_version(OPENSSL_VERSION));
                                                                         ^
```

nikic commented on ext/openssl/openssl.c:
```diff
@@ -705,6 +705,12 @@ static int X509_get_signature_nid(const X509 *x)
 
 #endif
 
+#define OpenSSL_version		OpenSSL_version
```
```suggestion
#define OpenSSL_version		SSLeay_version
```

Maybe?

neheb commented:
Yep. Will fix.

neheb commented:
Rebased against master. Also fixed a small issue with TLS1.3 support by testing that it's actually compiled in. OpenSSL 1.1 can be compiled without TLS 1.3 support

nikic commented:
Travis build failure due to redeclared macros:

```
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c:109:0: error: "PHP_OPENSSL_MAX_PROTO_VERSION" redefined [-Werror]
 #define PHP_OPENSSL_MAX_PROTO_VERSION STREAM_CRYPTO_METHOD_TLSv1_3
 ^
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c:69:0: note: this is the location of the previous definition
 #define PHP_OPENSSL_MAX_PROTO_VERSION STREAM_CRYPTO_METHOD_TLSv1_2
 ^
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c: In function ‘php_openssl_get_crypto_method_ctx_flags’:
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c:1030:22: error: ‘SSL_OP_NO_TLSv1_3’ undeclared (first use in this function)
   ssl_ctx_options |= SSL_OP_NO_TLSv1_3;
                      ^
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c:1030:22: note: each undeclared identifier is reported only once for each function it appears in
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c: In function ‘php_openssl_capture_session_meta’:
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c:1828:8: error: ‘TLS1_3_VERSION’ undeclared (first use in this function)
   case TLS1_3_VERSION:
        ^
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c: In function ‘php_openssl_sockop_set_option’:
/home/travis/build/php/php-src/ext/openssl/xp_ssl.c:2437:11: error: ‘TLS1_3_VERSION’ undeclared (first use in this function)
      case TLS1_3_VERSION: proto_str = "TLSv1.3"; break;
```

neheb commented:
Compiles now. Was a bad rebase.

nikic commented on ext/ftp/php_ftp.c:
```diff
@@ -318,12 +318,14 @@ static void ftp_destructor_ftpbuf(zend_resource *rsrc)
 PHP_MINIT_FUNCTION(ftp)
 {
 #ifdef HAVE_FTP_SSL
+#if PHP_OPENSSL_API_VERSION < 0x10100
```
I don't think this constant is available here, as it's part of php_openssl.h, which is not included here (and I think also shouldn't be included for dependency reasons). This check should be done using vanilla openssl constants instead.

KalleZ replied:
The SSL code enabling code seems differs a lot from Unix to Windows; On Unix this seems to depend on `ext/openssl` being enabled, whereas on Windows it just depends the OpenSSL library. I suppose this was made in the effort to actually make `ftp_ssl_connect()` available on Windows.

Either way @nikic is correct; this needs another way of doing it, at least to make it work on Windows builds like `configure --disable-all --enable-cli --enable-ftp=shared`

neheb replied:
Fixed.

nikic commented on ext/openssl/xp_ssl.c:
```diff
@@ -60,9 +60,19 @@
-#if OPENSSL_VERSION_NUMBER >= 0x10101000
+#endif
+
+#if OPENSSL_VERSION_NUMBER >= 0x10101000 && !defined(OPENSSL_NO_TLS1_3)
```
The checks above use `NO_TLS1_x_METHOD`, while this one is just `NO_TLS1_3`. What's the difference between these two?

bukka commented on ext/openssl/openssl.c:
```diff
@@ -705,6 +705,12 @@ static int X509_get_signature_nid(const X509 *x)
 
 #endif
 
+#define OpenSSL_version		SSLeay_version
```
Doesn't this cause double definitions for configuration without `OPENSSL_API_COMPAT` (e.g. for config like: `./config shared`?

neheb replied:
I believe this code block is found within an ```#if``` directive. https://github.com/php/php-src/blob/master/ext/openssl/openssl.c#L582

bukka commented on ext/openssl/xp_ssl.c:
```diff
@@ -60,9 +60,19 @@
 #define STREAM_CRYPTO_METHOD_TLSv1_3       (1<<6)
 
+#ifndef OPENSSL_NO_TLS1_METHOD
+#define HAVE_TLS1 1
```
I think you might be missing more places to make this work. From the quick look, this might need to be also covered in `php_openssl_ssl_socket_factory`. But this really needs some testing. It's also unrelated to this PR so I would prefer to have `OPENSSL_NO_TLS...` in a new PR as it might need a bit more testing.

neheb replied:
It's not so much about making it work as it is about making it compile. Making this work can be covered in another PR as you said.

bukka commented:
Here is a link for the bug that was already created for this some time ago: https://bugs.php.net/bug.php?id=74378 . Some bits from it are already fixed and this might finish it hopefully.

nikic commented:
Merged as https://github.com/php/php-src/commit/32e6d08dcd25418c96d8fe2218802a07fdf3363d, sorry for the delay...
